src/Features/Core/Portable/NavigationBar/IRemoteNavigationBarItemService.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.NavigationBar
+{
+    internal interface IRemoteNavigationBarItemService
+    {
+        ValueTask<ImmutableArray<SerializableNavigationBarItem>> GetItemsAsync(
+            PinnedSolutionInfo solutionInfo, DocumentId documentId, bool supportsCodeGeneration, CancellationToken cancellationToken);
+    }
+
+    [DataContract]
+    internal sealed class SerializableNavigationBarItem
```

Comment:
flattened union type of all the different navbaritem types.  ugly, but this is how we serialize hierarchies elsewhere.

Reply:
I was going through commit-by-commit, and I think this is a bit ugly, but better than the split you had before, with the indexes spread across different files. That just seemed like a trap waiting to happen.